### ISSUE TITLE
UDP source port being re-written

### ISSUE THREAD

**Issue by JasonN3** [projectcalico/calico#5606]
## Expected Behavior
The source port should be consistent for UDP traffic to the pod

## Current Behavior
Currently the source port is being changed when being processed

## Steps to Reproduce (for bugs)
1. Cluster was installed using kubeadm. A /23 was used for pods, a /24 for services, and a /24 for external service. None of them overlap.
2. Calico was installed using the manifest install
3. Configure for BGP peering and disable full mesh
4. Enable eBPF and disable kube-proxy using the instructions at https://projectcalico.docs.tigera.io/maintenance/ebpf/enabling-bpf
5. Attempt to run any nslookup command within a busybox pod

## Context
This is blocking pod to pod communication as DNS lookups cannot be completed.

## Your Environment
* Calico version: 3.22.0
* Orchestrator version: kubernetes 1.23.3
* Operating System and version: Fedora CoreOS 35.20220116.3.0

Some diagnosing was done in slack: https://calicousers.slack.com/archives/CUKP5S64R/p1644544583561509

**Comment by JasonN3** [comment]
This is the debug log when running the nslookup
[debug.log](https://github.com/projectcalico/calico/files/8050850/debug.log)

**Comment by tomastigera** [comment]
```
enp1s0---I: New packet at ifindex=2; mark=0
enp1s0---I: No metadata is shared by XDP
enp1s0---I: IP id=9431 s=c0a80cb3 d=c0a80d99
enp1s0---I: UDP; ports: s=53 d=53751
enp1s0---I: CT-17 lookup from c0a80cb3:53
enp1s0---I: CT-17 lookup to   c0a80d99:53751
enp1s0---I: CT-17 Hit! NORMAL entry.
enp1s0---I: CT-17 First response packet? ifindex=2
enp1s0---I: CT-17 Disabling bypass to allow kernel RPF.
enp1s0---I: CT-17 result: 0x202
enp1s0---I: conntrack entry flags 0x0
enp1s0---I: CT Hit
enp1s0---I: IP id=9431 s=c0a80cb3 d=c0a80d99
enp1s0---I: Entering calico_tc_skb_accepted
enp1s0---I: src=c0a80cb3 dst=c0a80d99
enp1s0---I: post_nat=0:0
enp1s0---I: tun_ip=0
enp1s0---I: pol_rc=0
enp1s0---I: sport=53
enp1s0---I: flags=0
enp1s0---I: ct_rc=2
enp1s0---I: ct_related=0
enp1s0---I: ip->ttl 62
enp1s0---I: Traffic is towards host namespace, marking with 1000000.
enp1s0---I: Final result=ALLOW (0). Program execution time: 53974ns

the same packet but with a different source port

cali1037-I: New packet at ifindex=7; mark=1000000
cali1037-I: IP id=9431 s=c0a80cb3 d=c0a80d99
cali1037-I: UDP; ports: s=432 d=53751
cali1037-I: CT-17 lookup from c0a80cb3:432
cali1037-I: CT-17 lookup to   c0a80d99:53751
cali1037-I: CT-17 Miss.
cali1037-I: CT-17 result: NEW.
cali1037-I: conntrack entry flags 0x0
cali1037-I: IP id=9431 s=c0a80cb3 d=c0a80d99
cali1037-I: About to jump to policy program.
cali1037-I: Entering calico_tc_skb_accepted_entrypoint
cali1037-I: IP id=9431 s=c0a80cb3 d=c0a80d99
cali1037-I: Entering calico_tc_skb_accepted
cali1037-I: src=c0a80cb3 dst=c0a80d99
cali1037-I: post_nat=c0a80d99:53751
cali1037-I: tun_ip=0
cali1037-I: pol_rc=1
cali1037-I: sport=432
cali1037-I: flags=0
cali1037-I: ct_rc=0
cali1037-I: ct_related=0
cali1037-I: ip->ttl 61
cali1037-I: Allowed by policy: ACCEPT
cali1037-I: CT-ALL packet mark is: 0x1000000
cali1037-I: CT-ALL Asked to create entry but packet is marked as from another endpoint, doing lookup
cali1037-I: CT-ALL Creating tracking entry type 0 at 7396195108025.
cali1037-I: CT-ALL tracking entry flags 0x0
cali1037-I: NEW src_to_dst->ifindex 0
cali1037-I: CT-ALL Whitelisted dest side - to EP
cali1037-I: Final result=ALLOW (0). Program execution time: 82883ns
cali1037-E: New packet at ifindex=7; mark=0
cali1037-E: IP id=34065 s=c0a80d99 d=c0a80cb3
cali1037-E: ICMP; type=3 code=3
cali1037-E: CT-1 lookup from c0a80d99:0
cali1037-E: CT-1 lookup to   c0a80cb3:771
cali1037-E: IP id=34065 s=c0a80d99 d=c0a80cb3
cali1037-E: CT-ICMP: proto 17
cali1037-E: CT-1 related lookup from c0a80cb3:432
cali1037-E: CT-1 related lookup to   c0a80d99:53751
cali1037-E: CT-1 Hit! NORMAL entry.
cali1037-E: CT-1 First response packet? ifindex=7
cali1037-E: CT-1 result: 0x202
cali1037-E: CT-1 result: related
cali1037-E: conntrack entry flags 0x0
cali1037-E: CT Hit
cali1037-E: IP id=34065 s=c0a80d99 d=c0a80cb3
cali1037-E: Entering calico_tc_skb_accepted
cali1037-E: src=c0a80d99 dst=c0a80cb3
```

**Comment by fasaxc** [comment]
@JasonN3 are you running anything else that could touch the packets?  Another CNI?  IPVS?  It might be useful to see the full `iptables-save -c` output.  Perhaps we've got a clashing mark bit or something.

**Comment by tomastigera** [comment]
So as per [slack](https://calicousers.slack.com/archives/CUKP5S64R/p1644673373688039?thread_ts=1644544583.561509&cid=CUKP5S64R) there is this conntrack entry:

Here the wrong port is dport=169

```
udp      17 27 src=192.168.12.179 dst=192.168.13.153 sport=53 dport=54990 [UNREPLIED] src=192.168.13.153 dst=192.168.12.179 sport=54990 dport=169 mark=0 secctx=system_u:object_r:unlabeled_t:s0 use=1
```

Based on some rule, linux created itself such an entry. Bpf has no control over that. We need to figure out what created it.

**Comment by tomastigera** [comment]
In the provided debug.log the pod sent 2 packets to port 53 from 53751. Those packets bypassed iptables and were forwarded straight to the outbound interface. Only then we got a response and because of enforcing RPF (which we cannot do in bpf) we dropped the packet to iptables. That is the point where the weird conntrack was created and where the packet was modified. The could have been no conntrack entry prior to that. Maybe something outdate? Some leftover?
